Refactor IREECodegenAttrs to use typed array parameters

MacDue opened the pull request:
This reworks the attributes in IREECodegenAttrs so that typed [Optional]ArrayRefParameters are used instead of untyped ArrayAttrs.

Doing this removes the manual type-checking and manual conversions to/from SmallVector<T>.

This did require a small restructuring of the lowering_config attribute, so instead of having multiple nested arrays IntegerAttrs for tiling levels, it now has one array of TilingLevelAttrs. The TilingLevelAttr then contains the tile sizes and tile interchange values for a particular level.

The TilingLevelAttr has a custom assembly format that keeps its syntax as close as possible to the previous syntax, to avoid as much test churn as possible. We plan to follow up this patch with another change that extends the TilingLevelAttr to support scalable sizes (using the canonical MLIR syntax), which will need a custom parser anyway, so we don't see this as a dealbreaker.

Example configs:

```
// No tile interchange (no change in syntax)
#iree_codegen.lowering_config<tile_sizes = [[2, 64, 64, 64],
                                            [1, 1, 1, 4],
                                            [0, 0, 0, 0]]>
// With tile interchange
// Before:
#iree_codegen.lowering_config<
   tile_sizes = [[32, 64, 0], [8, 32, 0], [0, 0, 16]],
   tile_interchange = [[1, 0, 2], [], []]>

// After:
#iree_codegen.lowering_config<tile_sizes = [
   {sizes=[32, 64, 0], interchange=[1, 0, 2]}, [8, 32, 0], [0, 0, 16]]>
```

---

These changes makes the `lowering_config` easier to extend for a follow on patch that starts adding scalable tile sizes.

MacDue committed:
Refactor IREECodegenAttrs to use typed array parameters

This reworks the attributes in IREECodegenAttrs so that typed
[Optional]ArrayRefParameters are used instead of untyped ArrayAttrs.

Doing this removes the manual type-checking and manual conversions
to/from SmallVector<T>.

This did require a small restructuring of the lowering_config attribute,
so instead of having multiple nested arrays IntegerAttrs for tiling
levels, it now has one array of TilingLevelAttrs. The TilingLevelAttr
then contains the tile sizes and tile interchange values for a
particular level.

The TilingLevelAttr has a custom assembly format that keeps its syntax
as close as possible to the previous syntax, to avoid as much test
churn as possible. We plan to follow up this patch with another change
that extends the TilingLevelAttr to support scalable sizes (using the
canonical MLIR syntax), which will need a custom parser anyway, so we
don't see this as a dealbreaker.

Example configs:

```
  // No tile interchange
  // Before:
  #iree_codegen.lowering_config<tile_sizes    = [[2, 64, 64, 64],
                                                 [1, 1, 1, 4],
                                                 [0, 0, 0, 0]]>
  // After:
  #iree_codegen.lowering_config<tiling_levels = [[2, 64, 64, 64],
                                                 [1, 1, 1, 4],
                                                 [0, 0, 0, 0]]>

  // With tile interchange
  // Before:
  #iree_codegen.lowering_config<
   tile_sizes = [[32, 64, 0], [8, 32, 0], [0, 0, 16]],
   tile_interchange = [[1, 0, 2], [], []]>

  // After:
  #iree_codegen.lowering_config<tiling_levels = [
   {sizes=[32, 64, 0], interchange=[1, 0, 2]}, [8, 32, 0], [0, 0, 16]]>
```

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Dialect/IREECodegenAttrs.cpp:
```diff
@@ -126,6 +86,71 @@ LogicalResult TranslationInfoAttr::verify(
+//===----------------------------------------------------------------------===//
+
+void LoweringConfigTilingLevelAttr::print(mlir::AsmPrinter &printer) const {
+  auto tileInterchange = getInterchange();
```
I am curious why the assembly format parser/printer didnt work here?

MacDue replied:
Ideally you could write:
```
let assemblyFormat = [{
    (`{` `sizes` `=` `[` $sizes `]` `,` `interchange` `=` `[` $interchange^ `]` `}`)
    : (`[` $sizes `]`)?
}];
```
But unfortunately the `assemblyFormat` limits parameters to appearing once (even across groups), so this results in: `error: duplicate parameter 'sizes'`.

MaheshRavishankar replied:
I cant parse the `assemblyFormat` above, but are you trying to print/parse the `sizes` and `interchange` for each level at the same time?

MacDue replied:
With the new structure, the tiling levels are a list of `TilingLevelAttr`s, which contains both the sizes and interchange for a particular level. This makes the sizes and interchange 1D parameters to the `TilingLevelAttr`, which means they can be typed arrays (they have to be 1D as the typed arrays only support primitive values). 

The custom format is to make updating the tests easier (though I think looks pretty nice too :))

You could use the generated printer/parser here, but that'd only allow the format to be:
```
 #iree_codegen.lowering_config<tiling_levels = [
   {sizes=[32, 64, 0]}, {sizes=[8, 32, 0]}, {sizes=[0, 0, 16]}]>
```
The custom printer allows special-casing the "no tile interchange" case as:
```
 // Pretty much the same as the old syntax:
 #iree_codegen.lowering_config<tiling_levels = [
   [32, 64, 0], [8, 32, 0], [0, 0, 16]]>
```
This makes most tests much easier to update (just `s/tile_sizes/tiling_levels`). The full `{sizes=[32, 64, 0], interchange=[1, 0, 2]}` is only used in a handful of tests as most places don't use `tile_interchange`.  

Hope that makes sense!

MaheshRavishankar replied:
Thanks for the details! THat makes sense.

MaheshRavishankar reviewed (changes requested):
This is a really nice change! Mostly looks OK to me and a good cosmetic change.
This does have some downstream effect, so we might need to stage this a bit. So marking "request changes" for now, but as soon as I get signal on that, this looks good.

MacDue committed:
Update missed test cases and test generators

MacDue committed:
Fix python lint error

MacDue committed:
Call the levels `tile_sizes` rather than `tiling_levels`

This is not really accurate, but reduces the churn and may become true
again if the interchange is moved around.

MaheshRavishankar reviewed (approved):
Ok, approving to land. Please note that if this causes issues downstream we might revert this to allow for downstream users to catch up. I'll try to not do the revert if possible, but that is a lever we might use if things get very hairy downstream. So i'd request any future changes that rely on it to be on hold for a couple of weeks if possible. 

cc @monorimet @yzhang93
